core/go/internal/publictxmgr/transaction_orchestrator.go

```diff
@@ -220,6 +229,102 @@ func (oc *orchestrator) getFirstInFlight() (ift *inFlightTransactionStageControl
+		WithContext(ctx).
+		Where(`"from" = ?`, oc.signingAddress).
+		Where("nonce IS NOT NULL").
+		Order("nonce DESC").
+		Limit(1).
+		Find(&txns).
+		Error
+	if err != nil || len(txns) == 0 {
+		return err
+	}
+	nextNonce := *txns[0].Nonce + 1
+	oc.nextNonce = &nextNonce
+	log.L(ctx).Infof("Next nonce initialized from DB fro %s: %d", oc.signingAddress, nextNonce)
+	return nil
+}
+
+func (oc *orchestrator) allocateNonces(ctx context.Context, txns []*DBPublicTxn) error {
+
+	// Of the the transactions might have nonces already
+	toAlloc := make([]*DBPublicTxn, 0, len(txns))
+	for _, tx := range txns {
+		if tx.Nonce == nil {
+			toAlloc = append(toAlloc, tx)
+		}
+	}
+	if len(toAlloc) == 0 {
+		// Nothing to do
+		return nil
+	}
+
+	// We need to ensure we have the next nonce to allocate
+	if oc.nextNonce == nil || time.Since(oc.lastNonceAlloc) > oc.nonceCacheTimeout {
+		log.L(ctx).Debugf("no cached nonce, or nonce expired for %s (cached=%v)", oc.signingAddress, oc.lastNonceAlloc)
+		txCount, err := oc.ethClient.GetTransactionCount(ctx, oc.signingAddress)
+		if err != nil {
+			return err
+		}
+		// See if we have nonces in our DB that are ahead of the mempool.
+		if oc.nextNonce != nil && *oc.nextNonce >= txCount.Uint64() {
+			log.L(ctx).Infof("Next nonce for %s is %d (at or ahead of mempool %d)", oc.signingAddress, *oc.nextNonce, txCount.Uint64())
+		} else {
+			// Otherwise take the node's answer
+			oc.nextNonce = (*uint64)(txCount)
+			log.L(ctx).Infof("Next nonce for %s set to %d (from eth_getTransactionCount)", oc.signingAddress, *oc.nextNonce)
+		}
+	}
+
+	// Set up the list of nonces we'll allocated, but until it's in the DB we do NOT update the oc.nextNonce beyond the first in the list
+	newNextNonce := *oc.nextNonce
+	newNonces := make([]uint64, len(toAlloc))
+	for i := range newNonces {
+		newNonces[i] = newNextNonce
+		newNextNonce++
+	}
+
+	// Run the DB TXN using a VALUES temp table to update multiple rows in a single operation
+	err := oc.p.DB().Transaction(func(dbTX *gorm.DB) error {
+		sqlQuery := `WITH nonce_updates ("pub_txn_id", "nonce") AS ( VALUES `
+		values := make([]any, 0, len(toAlloc)*2)
+		for i, tx := range toAlloc {
+			if i > 0 {
+				sqlQuery += `, `
+			}
+			sqlQuery += `( CAST (? AS BIGINT), CAST (? AS BIGINT) ) `
+			values = append(values, tx.PublicTxnID)
+			values = append(values, newNonces[i])
+			log.L(ctx).Debugf("assigning %s:%d (pubTxnId=%d)", oc.signingAddress, newNonces[i], tx.PublicTxnID)
+		}
+		sqlQuery += ` ) UPDATE "public_txns" SET "nonce" = nu."nonce" FROM ( SELECT "pub_txn_id", "nonce" FROM nonce_updates ) AS nu ` +
+			`WHERE "public_txns"."pub_txn_id" = nu."pub_txn_id";`
+		return dbTX.WithContext(ctx).Exec(sqlQuery, values...).Error
+	})
+	if err != nil {
+		return err
+	}
+
+	// Update the txns themselves, and our nextNonce
+	for i, tx := range toAlloc {
+		nonce := newNonces[i]
+		tx.Nonce = &nonce
+	}
+	oc.lastNonceAlloc = time.Now()
+	oc.nextNonce = &newNextNonce
+
+	return nil
+}
```

Comment:
This is the most important part of the change - the synchronous code to ensure we assign nonces to all transactions, before we do any processing on them.

It took me quite a bit of work to determine how in PSQL/SQLite I could use a single SQL call to `UPATE` and set multiple values (as this is very hot code performance wise). Further optimization might be possible.